File: 2948-make-lexicographically-smallest-array-by-swapping-elements/2948-make-lexicographically-smallest-array-by-swapping-elements.go
import (
  "slices"
)

func lexicographicallySmallestArray(nums []int, limit int) []int {
  // split into backward sorted groups and build group index
  sorted := slices.Clone(nums)
  slices.SortFunc(sorted, func(a int, b int) int { return a - b })
  groups := [][]int{ []int{ sorted[0] } }
  index := map[int]int{ sorted[0]: 0 }
  for i := 1; i < len(sorted); i++ {
    j := len(groups) - 1
    if sorted[i] - groups[j][len(groups[j]) - 1] <= limit {
      groups[j] = append(groups[j], sorted[i])
    } else {
      groups = append(groups, []int{ sorted[i] })
    }
    index[sorted[i]] = len(groups) - 1
  }

  // replace elements in input array
  for i := 0; i < len(nums); i++ {
    j := index[nums[i]]
    nums[i] = groups[j][0]
    groups[j] = groups[j][1:]
  }

  return nums
}

File: 2948-make-lexicographically-smallest-array-by-swapping-elements/2948-make-lexicographically-smallest-array-by-swapping-elements.ts
// forbiddance of permutation of particular pairs breaks the logic of sorting
// function lexicographicallySmallestArray(nums: number[], limit: number): number[] {
//   nums.sort((a: number, b: number) => {
//     if (a === b || Math.abs(a - b) > limit) return 0;
//     return a - b
//   });
//   return nums;
// };

function lexicographicallySmallestArray(nums: number[], limit: number): number[] {
  // split into backward sorted groups and build group index
  const sorted = nums.slice();
  sorted.sort((a, b) => b - a);
  const groups: number[][] = [[sorted[0]]];
  const index = { [sorted[0]]: 0 }
  for (let i = 1; i < sorted.length; i++) {
    if (groups.at(-1).at(-1) - sorted[i] <= limit) {
      groups.at(-1).push(sorted[i]);
    } else {
      groups.push([sorted[i]]);
    }
    index[sorted[i]] = groups.length - 1
  }
  // replace elements in input array
  for (let i = 0; i < nums.length; i++) {
    nums[i] = groups[index[nums[i]]].pop()
  }
  return nums;
};